experiments/dom.html

```diff
@@ -0,0 +1,404 @@
+                html.push("<span class='", tokens[i], "'>", tokens[i], "</span>")
+            }
+        }
+        return update;
+    }
+    tests.domNewNodes = function() {
+        function elt(parent, name, className, content) {
+            var el = document.createElement(name);
+            if (content) {
+                el.appendChild(document.createTextNode(content))
+            }
+            parent && parent.appendChild(el)
+            el.className = className
+            return el;
+        }
+        function txtEl(parent, str) {
+            return parent.appendChild(document.createTextNode(str))
+        }
+        function clearEl(el) {
+            while (el.lastChild)
+                el.removeChild(el.lastChild);
+        }
+
+        
+        function update(rootEl, lines) {
+            clearEl(rootEl)
+            for (var i = 0; i < lines.length; i++) {
+                var el = elt(rootEl, "div", "line")
+                renderLine(el, lines[i])
+            }
+        }
+    
+        function renderLine(html, tokens) {
+            for (var i = 0; i < tokens.length; i ++) {
+                elt(html, "span", tokens[i], tokens[i])
+            }
+        }
+        return update;
+    }
+    tests.domCloneNode = function() {
+        var els = {}
+        function elt(parent, name, className, content) {
+            if (!els[name]) els[name] = document.createElement(name);
+            var el = els[name].cloneNode(false)
+            if (content) {
+                el.appendChild(document.createTextNode(content))
+            }
+            parent && parent.appendChild(el)
+            el.className = className
+            return el;
+        }
+        function txtEl(parent, str) {
+            return parent.appendChild(document.createTextNode(str))
+        }
+        function clearEl(el) {
+            while (el.lastChild)
+                el.removeChild(el.lastChild);
+        }
+
+        
+        function update(rootEl, lines) {
+            clearEl(rootEl)
+            for (var i = 0; i < lines.length; i++) {
+                var el = elt(rootEl, "div", "line")
+                renderLine(el, lines[i])
+            }
+        }
+    
+        function renderLine(html, tokens) {
+            for (var i = 0; i < tokens.length; i ++) {
+                elt(html, "span", tokens[i], tokens[i])
+            }
+        }
+        return update;
+    }
+    tests.domReuseNodes = function() {
+        function elt(parent, name, className, content) {
+            var el = document.createElement(name);
+            if (content) {
+                el.appendChild(document.createTextNode(content))
+            }
+            parent && parent.appendChild(el)
+            el.className = className
+            return el;
+        }
+        function txtEl(parent, str) {
+            return parent.appendChild(document.createTextNode(str))
+        }
+        function clearEl(el, l) {
+            while (l --> 0)
```

Comment:
Very cute but are you sure we should use that trick? Sometimes smart code like this triggers questions or causes people to make the wrong assumptions.

Reply:
Well this is in experiments, so using the trick is not very harmful, but i'll remove this later, when i add one more drawing method to this.